### PR TITLE
Yul documentation

### PR DESCRIPTION
**Pull request by LucasSte**
This PR adds documentation for Yul. Changes in `docs/conf.py` were necessary to support a new linter for both Solidity and Yul.

### CODE REVIEW

**Comment by seanyoung** [comment]
With this PR, yul is working so the top level README.md and CHANGELOG.md can be updated

**Comment by seanyoung** [comment]
With this PR, yul is working so the top level README.md and CHANGELOG.md can be updated

File: docs/yul_language/builtins.rst [hunk 1]
```diff
@@ -0,0 +1,202 @@
+Builtins
+========
+
+Most operations in Yul are performed via builtin functions. Solang supports
+most builtins, however memory operations and chain operations are not yet implemented.
+They will be added in a future release.
```
Comment (seanyoung) [inline docs/yul_language/builtins.rst:6]: This section should NOT say "they will be added", because many will never be added. We should add something about the origin of the builtins: 

yul builtins are low level EVM instructions and many are ethereum specific (link to https://ethereum.org/en/developers/docs/evm/opcodes/). On Solana or Substrate these instructions do not exist, and because of low-level differences between the chains, it is impossible to implement some builtins. For example ``delegatecall`` and ``staticcall`` just does not exist as a concept on either chain.

File: docs/yul_language/functions.rst [hunk 2]
```diff
@@ -0,0 +1,31 @@
+Functions
+=========
+
+Functions in Yul cannot access any variable outside their scope, i.e. they can
+only operate with the variables they receive as arguments. You can define types for arguments and returns. If no
+type is specified, the compiler will default to ``u256``.
```
Comment (seanyoung) [inline docs/yul_language/functions.rst:6]: This is a little terse.

File: docs/yul_language/functions.rst [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            ret1 := mulmod(c, d, e)
+            ret2 := addmod(c, d, e)
+        }
+
+        function noReturns(a, b)
+        {
+            store(a, b)
+        }
+    }
```
Comment (seanyoung) [inline docs/yul_language/functions.rst:31]: This example does not show how it would be used in solidity or how to do function calls.

File: docs/yul_language/statements.rst [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            funcCall(4, 3)
+        } // There cannot be an 'else' here
+    }
+
+
+Switch
+_______
+
+Switch statements are not yet supported in Solang. If there is urgent need to support them,
+please, file a GitHub issue in the official repository.
```
Comment (seanyoung) [inline docs/yul_language/statements.rst:46]: no need for official

File: docs/yul_language/statements.rst [hunk 5]
```diff
@@ -0,0 +1,118 @@
+Statements
+==========
+
+For-loop
+________
+
+For-loops are completely supported in Solang. ``continue`` and ``break`` statements are also supported.
```
Comment (seanyoung) [inline docs/yul_language/statements.rst:7]: some words about how for loops are written would be helpful (the init, cond, next block/expression)

File: docs/yul_language/statements.rst [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        let j : u32, k : u8 := manyReturns()
+    }
+
+Assignments
+___________
+
+Variables can be assignment using the ``:=`` operator. If the types do not match,
+the compiler performs an implicit cast, so that the right hand side type matches that of the variable.
+Multiple variables can be assigned in a single line if the right hand side is a function call that returns multiple
+values.
```
Comment (seanyoung) [inline docs/yul_language/statements.rst:93]: Does the compiler ever produce an error/warning?

File: docs/yul_language/yul.rst [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+.. _yul_section:
+
+Yul, also know as EVM assembly, is a low level language for creating smart contracts and for providing more
+control over the execution environment when using Solidity as the primary language. Although it enables
+more possibilities to manage memory, using Yul does not imply a performance improvement. In Solang,
+all Yul constructs are processed using the same pipeline as Solidity.
+
+The support for Yul is only partial. We support all statements, except for the switch-block. In addition,
+some Yul builtins are not yet implemented, but we are trying to improve compatibility. In the following sections,
```
Comment (seanyoung) [inline docs/yul_language/yul.rst:12]: do not give the impression that all builtins will be implemented.

remove `, but we are trying to improve compatibility`, this is implied and rather than focus on what we haven't done, it's better to focus on what we have done.

**Review by seanyoung — CHANGES_REQUESTED** [review]

File: docs/yul_language/statements.rst (hunk 6 above)
Comment (LucasSte) [reply docs/yul_language/statements.rst:93]: Yes, it does

File: docs/yul_language/statements.rst [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+
+Variable declaration
+____________________
+
+Variables can be declared in Yul using the `let` keyword. Multiple variables can be declared at the same line
+if there is no initializer or the initializer is a functon that returns multiple values.
+
+The default type for variables in Yul is ``u256``. If you want to declare a variable with another type, use the colon.
+Note that if the variable type and the type of the right hand side of the assignment disagree, there will be an implicit
```
Comment (seanyoung) [inline docs/yul_language/statements.rst:74]: replace disagree with "are different"

Comment (seanyoung) [reply docs/yul_language/statements.rst:74]: or "do not match"

File: docs/yul_language/statements.rst (hunk 6 above)
Comment (seanyoung) [reply docs/yul_language/statements.rst:93]: I worry that "implicit cast" won't be understood. How about "converted to the right type, if possible".

File: docs/yul_language/functions.rst (hunk 2 above)
Comment (LucasSte) [reply docs/yul_language/functions.rst:6]: I didn't want to make it prolix, but I'll improve it.

File: docs/yul_language/functions.rst (hunk 3 above)
Comment (LucasSte) [reply docs/yul_language/functions.rst:31]: This section is about Yul, not inline assembly, so I should not mention Solidity here. There interaction of Yul and Solidity is described at `inline_assembly.rst`.

File: docs/language/inline_assembly.rst [hunk 9]
```diff
@@ -0,0 +1,116 @@
+Inline Assembly
+===============
+
+In Solidity functions, developers are allowed to write assembly blocks containing Yul code. For more information about
+the Yul programming language, please refer to the :ref:`yul section <yul_section>`.
+
+In an assembly block, you can access solidity local variables freely and modify them as well. Bear in mind, however,
+that strings, vectors and structs are reference types, so accessing them in Yul will return a memory address. Any
```
Comment (seanyoung) [inline docs/language/inline_assembly.rst:8]: I would rephrase this as "reference types like string, .. are memory addresses in yul, so manipulating them can be unsafe unless done correctly"

**Review by seanyoung — APPROVED** [review]
Looks great!
